hibernate/cache-v53/src/main/java/org/infinispan/hibernate/cache/v53/impl/FutureUpdateInvocation.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.infinispan.hibernate.cache.v53.impl;
+
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import org.infinispan.functional.FunctionalMap;
+import org.infinispan.hibernate.cache.commons.InfinispanDataRegion;
+import org.infinispan.hibernate.cache.commons.util.FutureUpdate;
+import org.infinispan.hibernate.cache.commons.util.InfinispanMessageLogger;
+
+/**
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+public class FutureUpdateInvocation implements Invocation, Function<Throwable, Void> {
+   private static final InfinispanMessageLogger log = InfinispanMessageLogger.Provider.getLog( FutureUpdateInvocation.class );
+
+   private final UUID uuid = UUID.randomUUID();
+   private final Object key;
+   private final Object value;
+   private final InfinispanDataRegion region;
+   private final long sessionTimestamp;
+   private final FunctionalMap.ReadWriteMap<Object, Object> rwMap;
+
+   public FutureUpdateInvocation(FunctionalMap.ReadWriteMap<Object, Object> rwMap,
+                                 Object key, Object value, InfinispanDataRegion region, long sessionTimestamp) {
+
+      this.rwMap = rwMap;
+      this.key = key;
+      this.value = value;
+      this.region = region;
+      this.sessionTimestamp = sessionTimestamp;
+   }
+
+   public UUID getUuid() {
+      return uuid;
+   }
+
+   @Override
+   public CompletableFuture<Void> invoke(boolean success) {
+      // If the region was invalidated during this session, we can't know that the value we're inserting is valid
+      // so we'll just null the tombstone
+      if (sessionTimestamp < region.getLastRegionInvalidation()) {
+         success = false;
+      }
+      // Exceptions in #afterCompletion() are silently ignored, since the transaction
+      // is already committed in DB. However we must not return until we update the cache.
+      FutureUpdate futureUpdate = new FutureUpdate(uuid, region.nextTimestamp(), success ? this.value : null);
+      for (;;) {
```

Comment:
Could this get noisy? Couldn't we just remove the remove the entry from the cache if we have any issues (and log those silently) rather than trying until it works?

Reply:
A removal is as likely to fail as any other update.

Reply:
Right.